ouroboros-consensus/src/Ouroboros/Consensus/Protocol/PBFT.hs

```diff
@@ -156,31 +157,45 @@ instance (PBftCrypto c, Typeable c) => OuroborosTag (PBft c) where
     where
       PBftParams{..}  = pbftParams
 
-  applyChainState toEnc cfg@PBftNodeConfig{..} (PBftLedgerView dms) b (signers, lastSlot) = do
+  applyChainState toEnc cfg@PBftNodeConfig{..} lv@(PBftLedgerView dms) b chainState = do
       -- Check that the issuer signature verifies, and that it's a delegate of a
       -- genesis key, and that genesis key hasn't voted too many times.
-
-      unless (verifySignedDSIGN toEnc (pbftIssuer payload)
-                      (blockPreHeader b)
-                      (pbftSignature payload))
-        $ throwError PBftInvalidSignature
+      case verifySignedDSIGN
+             toEnc
+             (pbftIssuer payload)
+             (blockPreHeader b)
+             (pbftSignature payload) of
+        Right () -> return ()
+        Left err -> throwError $ PBftInvalidSignature err
+
+      let (signers, lastSlot) = ( takeR winSize $ fst <$> chainState
+                                , maybe (SlotNo 0) snd $ Seq.lookup (Seq.length chainState) chainState
+                                )
 
       unless (blockSlot b > lastSlot)
         $ throwError PBftInvalidSlot
 
-      case Map.lookup (pbftIssuer payload) $ invertBijection dms of
-        Nothing -> throwError PBftNotGenesisDelegate
+      case Bimap.lookup (hashVerKey $ pbftIssuer payload) $ Bimap.twist dms of
```

Comment:
Maybe a reverse lookup rather than a lookup after "twist" would be faster..? (Minor.)